src/fluree/db/util/async.cljc

```diff
@@ -105,3 +108,57 @@
   "Returns true if core async channel."
   [x]
   (satisfies? async-protocols/Channel x))
+
+;;  "OpenTelemetry uses ThreadLocal storage to propagate trace state.
+;;   This typically requires wrapping your tasks or executors so that this state
+;;   is propagated to threads when desired.
+;;   See https://github.com/open-telemetry/opentelemetry-java/blob/main/context/src/main/java/io/opentelemetry/context/Context.java
+;;   core.async does not expose the thread pool direclty so we need to monkey patch it.
+
+;;   core.async will propagate clojure var bindings, so alternatively it might be possible
+;;   to implement a ContextStorageProvider that stores state in a var instead of thread locals.
+;;   I spent a bit of time trying to figure this out but it seems quite a bit more involved than
+;;   patching core.aysnc"
+#?(:clj
+   (do
+     (defn wrap-mdc ^Runnable [^Runnable r]
+       (let [context (MDC/getCopyOfContextMap)]
+         (reify Runnable
+           (run [_]
+             (let [current (MDC/getCopyOfContextMap)]
+               (try
+                 (MDC/setContextMap context)
+                 (.run r)
+                 (finally
+                   (MDC/setContextMap current))))))))
+
+     (defonce original-run dispatch/run)
+
+     ;; if OpenTelemetry is on classpath than propagate otel context otherwise no-op
+     (def wrap-otel
+       (try
+         (let [context-class (Class/forName "io.opentelemetry.context.Context")
+               current-method (.getMethod context-class "current" (into-array Class []))
+               wrap-method (.getMethod context-class "wrap" (into-array Class [Runnable]))]
+           (fn wrap-otel ^Runnable [^Runnable r]
+             (let [current-context (.invoke current-method nil (into-array Object []))]
+               (.invoke wrap-method current-context (into-array Object [r])))))
+         (catch ClassNotFoundException _
+           (fn wrap-otel ^Runnable [^Runnable r] r))))
+
+     (defn wrapped-runable ^Runnable [^Runnable r] (wrap-otel (wrap-mdc r)))
+     (defn patched-run
+       "Runs Runnable r with OpenTelemetry context propagation."
+       [^Runnable r]
+       (original-run (wrapped-runable r)))
+
+     (alter-var-root #'dispatch/run (constantly patched-run))
```

Comment:
I get really squeamish about patching external function implementations we don't control, especially non-public implementation internals. Not only can it lead to subtle bugs that are difficult to diagnose, but it also relies on a non-contracted implementation that could change at any time either breaking everything all together or, worse, leading to undefined behavior and more difficult to diagnose subtle bugs. 

I think this is especially true in this case because `clojure.core.async` is about to totally change its implementation to move from their own async thread pool to using JVM native virtual threads provided by [project loom](https://openjdk.org/projects/loom/). That new implementation is already in alpha, and we plan on switching to it as soon as it's marked stable.